### ISSUE TITLE
BlAnnulusSectorGeometry don't display the same shape on Alexandrie / Sparta

### ISSUE THREAD

**Issue by Nyan11** [pharo-graphics/Bloc#187]
In a playground:

# 1 with Sparta back-end
```st
sectorGeometry := BlAnnulusSectorGeometry new 
		innerRadius: 0;
		outerRadius: 100;
		startAngle: 180;
		endAngle: 120;
		yourself.			

element := BlElement new
		geometry: sectorGeometry;
		background: Color blue;
		yourself.
	
space := BlSpace new useSpartaOSWindowSDL2Host.
space addChild: element.
space show.
```
![image](https://user-images.githubusercontent.com/34318678/199054371-7d647122-3dbe-4687-8cdb-27a9e7138146.png)

# 2 with Alexandrie back-end

```st
sectorGeometry := BlAnnulusSectorGeometry new
		innerRadius: 0;
		outerRadius: 100;
		startAngle: 180;
		endAngle: 120;
		yourself.			

element := BlElement new
		geometry: sectorGeometry;
		background: Color blue;
		yourself.
	
space := BlSpace new useAlexandrieOSWindowSDL2Host .
space addChild: element.
space show.
```

![image](https://user-images.githubusercontent.com/34318678/199054711-2b8d4624-3b04-4760-8aad-5639221cbe37.png)

**Comment by tinchodias** [comment]
Hello, this rendering diff was deliberate but not well documented... it is tested in one of BARenderTest example canvas.

Do you consider one of them is the correct? I have an opinion but don't want to bias yours.

thanks for reporting an issue.

**Comment by Nyan11** [comment]
Hello,

I prefer the idea of **Alexandrie** where you color the angle rather than make it disappear.

**Comment by Nyan11** [comment]
Also, the direction doesn't feel right. I feel like the angle should be choose in the counter-clock direction.

![image](https://user-images.githubusercontent.com/34318678/199536639-20d975a4-5fca-4ebf-9c69-c0fd54557dfd.png)

![image](https://user-images.githubusercontent.com/34318678/199537228-118cf8d8-72fe-47e3-9812-2fd15658a2fe.png)

![image](https://user-images.githubusercontent.com/34318678/199537741-6f8e7419-7d57-4f3b-896b-114b2b4946b1.png)

**Comment by labordep** [comment]
Hello @tinchodias :)
@Nyan11 is working with us, I have the same opinion of him. Regardless of the backend, the user should have the same drawing shape result, specialy for theses basics geometrics functions. Now there is a choice of a angle direction and in my opinion the trigonometric circle is a standard wich is good.

**Comment by tinchodias** [comment]
Hi @Nyan11 and @labordep 

I agree all alternative canvas should share the criteria to render this geometry... as it happens with other geometries (I think it's the only exception). I can do the change.

I add to the discussion what Gregory said in the lse-bloc mailing-list, a couple of months ago:

> My preference would be to use radians, counterclockwise, starting at point (1,0) on the unit circle.   Radians are good for representing periodicity, and counterclockwise projection would map most efficiently to ordinary formulaic results.  On the other hand, orientation on a graphic canvas, where the bottom right quadrant is drawn upon, may lend itself to clockwise.  I am not sure.

When I had to implement this geometry for Alexandrie, I observed how it behave in Sparta-Cairo and Sparta-Skia, and to my surprise they were different! It is buggy in Sparta-Cairo (I think when inner radius is not zero). And for Sparta-Skia, I didn't agree in how start and end angles are interpreted. 

So, as there was already a difference, I added my own: do it like in the Cairo API (e.g. to draw arcs). The angles are specified clockwise and starting at "three o'clock" for our eyes (I say it in this way, I think you will understand what I mean... I interpret it's what Gregory means with "starting at point (1,0) on the unit circle"). This is consistent with the rest of the API where, for our eyes, the (0,0) origin is at top-left and the (1,1) is at bottom-right respect of origin. In comparison to how we read a pie chart in a book, or how we'd draw angles in a piece of paper, this is vertically flipped... but it's the standard way in computer graphics as far as I know. In this context, I consider consistent that angles in computer grow clockwise, because this is "vertically flipped" as for positions.

Don't know if I am clear, sorry if not.

**Comment by labordep** [comment]
Hi @tinchodias, 
This is very clear, no problem for me, there is a logical and this is simple to understand. 
The most important is to have the same behavior between backend (when it is possible).

Thanks in advance for the patch :)

**Comment by tinchodias** [comment]
You will have news very soon! 
Note about the original snippets: the innerRadius and outerRadius must be in the range [0..1], this is said in the setters (aka "mutators").

**Comment by tinchodias** [comment]
@Nyan11 @labordep please tell me if you can load the change in the Pull Request #190 

Or maybe you prefer that I merge the change and test it on the trunk.

**Comment by fool-in-the-rain** [comment]
Hi, this is Greg, of the Annulus Sector argument. : ) 

Yes, I have converted all my code over to clockwise.   The thing is, I like it this way now.  It's still counterintuitive to me, but it's just as counterintuitive to me to think of points below the X axis as positive.    Pulling it all together so they both flip signs along the X axis is just a matter of changing the way you cross you street in London.

**Comment by fool-in-the-rain** [comment]
This is Gregory, of the Annulus Sector Argument above.   Having degrees negative below the X axis is consistent, in my mind now, with having points below the X axis positive y.   Both those ideas are counterintuitive, but they are consistent.  I think keeping them that way will pay dividends in other ways later.   A good comment or two in the code will help people understand.

**Comment by fool-in-the-rain** [comment]
I (Gregory) agree with the comment above about flipping vertically, around the X axis.  I.e. the unintuitive approach for the annulus sector's clockwise geometry is strangely consistent with the unintuitive way that graphics take place below the X axis.

I am moving from Skia in Pharo 9 now, to Alexandrie in Pharo 11.  Can I expect that the annulus sectors will continue in a clockwise direction, or are you planning on flipping it back to conventional geometry, counterclockwise rotation?

The major win I think would be using Radians.  The divisibility with radians is more natural.   It's more expressive.

**Comment by tinchodias** [comment]
@fool-in-the-rain 
So we agree in using clockwise.
I will add a radians API to the geometry.

**Comment by tinchodias** [comment]
These decisions must be specified at least in the geometry's class comment.

For the record, two references:

The [Cairo API](https://www.cairographics.org/manual/cairo-Paths.html#cairo-arc) specifies: "An angle of 0.0 is in the direction of the positive X axis (in user space). An angle of PI/2.0 radians (90 degrees) is in the direction of the positive Y axis (in user space). Angles increase in the direction from the positive X axis toward the positive Y axis.".

[This website](https://fcs-cs.github.io/cs1-2018/modules/01-introduction/computer-coordinates/) explains angles in computer graphics in similar way.

**Comment by tinchodias** [comment]
I commented about this issue in the blog post: https://pharoweekly.wordpress.com/2022/11/14/bloc-update/

**Comment by fool-in-the-rain** [comment]
That was excellent source material.  Confirmation with clarity.

**Comment by fool-in-the-rain** [comment]
The only other thing I would say, is it might be more convenient, in terms of how code actually gets written, to have the starting angle, and then the difference, that would make up the new angle.  It would clean up the development code, and push the algebra back into the math classes.   As a matter of fact, I'm pretty sure that's how it was before, when I was working in Pharo 7.  It was a starting angle, and a delta.

**Comment by tinchodias** [comment]
Thanks @fool-in-the-rain I like the approach of setting a delta to define the sector. I think I read that approach in some other library but don't remember well which one. Maybe the SVG standard, or Skia.
I was preparing a new PR related to this issue, but also (trying to keep the API backwards compatible) add the start and end angle radians mutator methods. I can add a endDeltaAngle: easily...

**Comment by fool-in-the-rain** [comment]
Agreed, that retaining the (backward compat)ability of specifying annulus sector via start/endpoint angles will prove helpful, for example, when processing columns of raw data.

**Comment by fool-in-the-rain** [comment]
For what it's worth, I'm experimenting with Fontlab today, which draws vectors, and I notice that it does something that I think we had with Skia: the origin is at 12 o'clock.   In other respects, it follows clockwise = positive.

**Comment by tinchodias** [comment]
Fontlab looks nice but never used it. 
As another reference, AFAIR, Inkscape UI handles it as I proposed in the PR.
I think that an UI or any other "higher-level layer" on top of Bloc might use another criteria for angles, and translate it the the angles that the low-level API (bloc) needs.

**Comment by fool-in-the-rain** [comment]
Absolutely agreed that the 'higher-level layer' should have its own independent way of charting space and translate it to whatever the supporting layer (Bloc in this case) requires as an API.  That's what I do.

**Comment by tinchodias** [comment]
What do you think of https://github.com/pharo-graphics/Bloc/pull/192#discussion_r1036634951 ?
